### PR TITLE
File conn supports indexing

### PR DESCRIPTION
**Pull request by bplatz**
This adds indexing support to the file connection.

It also addresses sending file data back upstream for use by consensus.

Garbage collection still needs to get analyzed, that will get done in a separate ticket. It likely needs a few adjustments, but didn't think it was worth holding up not having any indexing.

### CODE REVIEW

File: src/fluree/db/conn/file.cljc [hunk 1]
```diff
@@ -205,6 +211,12 @@
   (-ctx-read [conn context-key] (go (read-context conn context-key)))
   (-ctx-write [conn ledger context-data] (go (write-context conn ledger
                                                             context-data)))
+  (-index-file-write [conn ledger index-type index-data]
```
Comment (zonotope) [inline src/fluree/db/conn/file.cljc:214]: This should probably wait for a later pull request, but I think we should start moving to much slimmer protocols with more primitive methods to do basic operations like accessing storage, and implementing higher level functionality in terms of those methods using plain functions that accept protocol implementations as arguments. I prefer to use protocols to wrap primitive, external interactions with stateful resources like an s3 bucket, file system, or some kind of external api. 

To make that concrete, in this case, I think we should have generic read and write methods on a storage record, and, if necessary, have different storage records attached to a connection for things like contexts and indexes, but only if those things need to be stored in different places. 

I'd like to keep protocols as close to the "metal" (or cloud, or api request, whatever) as possible, and implement our business logic functionality in term of those primitives. That will make the the protocols more general, less concerned with changing implementation details, and more composable.

File: src/fluree/db/storage/core.cljc [hunk 2]
```diff
@@ -1,198 +1,99 @@
...
+      (assoc branch :id (:address res)))))
 
 (defn write-garbage
   "Writes garbage record out for latest index."
   [db garbage]
   (go-try
-    (let [{:keys [conn network ledger-id block]} db
-          garbage-key (ledger-garbage-key network ledger-id block)
+    (let [{:keys [conn ledger ledger-id t]} db
+          t'          (- t) ;; use positive t integer
```
Comment (zonotope) [inline src/fluree/db/storage/core.cljc:79]: Also not for this pr, but I think we should think about handling this conversion (and possibly others) in a more systematic way at the boundaries to make it less likely someone will for get to (de)apply it when dealing with `t` values.

**Review by zonotope — APPROVED** [review]
I have a few suggestions that should probably be addressed later. This looks good to me otherwise.

File: src/fluree/db/storage/core.cljc (hunk 2 above)
Comment (bplatz) [reply src/fluree/db/storage/core.cljc:79]: Perhaps it should be applied in the serialization / deserialization protocol.

**Pull request merged**